### PR TITLE
[WIP] Regulated Assets Draft

### LINKED ISSUE #135: Allow asset issuers to require authorization per transaction
# Motivation
There is a growing number of requirements, set by asset issuers and regulators, regarding assets and trust-lines. From velocity limits to minimal balances and weird fee structures. (Related issues:  #101, #85, #71)

# Introduction
Instead of supporting all of these different requirements, we could delegate this logic to an off-chain service.  

*Note: this is not a complete proposal. just a starting point for discussion.* 

I propose adding the ability to declare that an asset is **TX_AUTHORIZATION_REQUIRED**. Meaning that this asset requires the issuer's signature for every transaction involving the asset. The issuer will commit to providing a signing service, that signs compliant transactions and provides meaningful errors for non-compliant ones.

I propose two approaches for achieving this:
- Approach 1 (preferred): combining a protocol change and an ecosystem proposal. 
- Approach 2: no protocol changes. only an ecosystem proposal + making the issuer a co-signer on accounts that hold their trustline. 

Given that both approaches utilize a similar ecosystem proposal, and that protocol changes take time, we can start from 2 and then easily migrate to 1 if/when the protocol changes. 

# Approach 1 
## Protocol changes
- Allow TX_AUTHORIZATION_REQUIRED flag on accounts.
- Require issuer signature for every transaction involving an asset that has TX_AUTHORIZATION_REQUIRED

**UPDATE: a possible outline for the protocol changes: https://github.com/stellar/stellar-protocol/issues/146**

## Ecosystem Proposal
- Add a TX_AUTH_SERVER to asset issuer's TOML. This points to their off-chain signing service. 
- Add a TX_AUTH_REQUIREMENTS to asset issuer's TOML. This will outline which specific requirements the issuer is looking for in transactions. This can be a human readable string or a transaction template or a set of predefined requirements that wallets can comply with. (TBD)
- Define the protocol for a signer service. Essentially, a simple HTTP service that receives transactions. It signs compliant ones and provides meaningful errors for non-compliant ones.

# Approach 2
Without any changes to the protocol, the issuers can create accounts for their signers, in which they are a co signer (2/2 scheme). This allows the SEP from approach 1 to be used as is. 
However, this presents some issues, to name a few: 
- The account will need issuer approval for transactions unrelated to the issuer (XLM transfers, etc.) 
- Having multiple trustlines for assets that require tx authorization will put us in a world of pain.

### PR DESCRIPTION
**Pull request by tomerweller**
This is an initial draft for a "Regulated Assets" protocol (closes #135) 
Easy reading here: https://github.com/tomerweller/stellar-protocol/blob/regulated-assets/drafts/draft-0003.md

Would appreciate any feedback.

### CODE REVIEW

**Comment by michaeldowling** [comment]
Hey @tomerweller - this looks mostly good, though now that I am reading through this and thinking about it, I'm not sure why there's a validation service.  Or, more specifically, not sure why validation and signing are 2 separate services.

I get that validation can be used for verifying well-formed transactions (i.e. to ensure you send X of your asset Y to party Z if you use asset Q), but wouldn't that also require a signature?  And if so, why not combine with the approval service, and have a single endpoint for validation, and if valid, also sign from the issuer?

File: drafts/draft-0004.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+Rejected (status code 400)
+A rejection response means that the transaction is not compliant and could not be made compliant.
+
+Parameters:
+Name
+Data Type
+Description
+error
+string
+A human readable string explaining why the transaction is not compliant and could not be made compliant.
```
Comment (michaeldowling) [inline drafts/draft-0004.md]: Keep in mind that most of the time, regulated agencies require that the reasons for rejections of funds transfers are for the financial institution itself, but not for consumption by the account holder.  This is for security and other reasons.  So, if you fail a compliance check, the "reason" returned from here would probably be "You should contact your issuer."

File: drafts/draft-0004.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+Request 
+HTTP POST request is encoded as application/json.
+
+Parameters :
+Name
+Data Type
+Description
+tx
+string
+Transaction XDR, Base64 encoded. Signed by account holder/s
```
Comment (michaeldowling) [inline drafts/draft-0004.md]: Currently, I don't care if the account holder sends me a signed TX for issuer signature, or an unsigned TX.  Either way, the TX can't be submitted without both.

Can this be an unsigned or signed transaction?

File: drafts/draft-0004.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+A human readable string explaining why the transaction is not compliant and could not be made compliant.
+Best Practices
+Once a transaction has been signed, the issuer should mark it as pending and take it into account, as long as it hasn’t been timed out, so that they can have a accurate view of the world. 
+Account Setup
+Implementing Regulated Assets requires the participating accounts to be “managed” by the issuer. This is achieved by having the issuer a co-signer on the account, such that a medium threshold operation cannot be submitted without the issuer’s approval. 
+
+In the future, this requirement can be replaced by introducing protocol level CoSigned assets.
+Discussion
+Should my asset be a regulated asset ? 
+Ideally, No. Implementing Regulated Assets should be a last resort. It comes with a substantial operational overhead, added complexity, and burden for the users. Issuers should only go down this route if it absolutely required.
```
Comment (michaeldowling) [inline drafts/draft-0004.md]: It's the only resort.

**Comment by michaeldowling** [comment]
Okay, duh, I just read the last couple sentences.  So, validation service might augment the original TX with additional operations or changes.

Interesting.  I see how that could be very helpful!  Our service won't use it I don't think but I see the utility now.

File: drafts/draft-0003.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+-----|-----------|------------
+tx|string|Unsigned Transaction XDR, Base64 encoded. This transaction will be tested for compliance.
+
+### Responses
+
+*HTTP responses are encoded as application/json*
+
+#### Success (*status code 200*)
+
+A successful response means that the transaction is compliant or the validation service was able to modify it to meet compliance.
```
[the remark this answers is not in the document]
Comment (michaeldowling) [reply drafts/draft-0003.md]: ^^ im in agreement, though I think you can also make a choice as to whether you as an anchor need or want it, yes?

[the remark this answers is not in the document]
Comment (tomerweller) [reply drafts/draft-0003.md]: Maybe add a boolean flag in the toml section that indicates whether the validation server will augment transactions?

File: drafts/draft-0003.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+### Best practices
+
+- ALWAYS explain the changes that were made to a transaction through the message parameter.
+- Core operations shouldn't be modified as that can be confusing and misleading. For example, if the user wishes to put an offer for 1000 GOATS but due to velocity limits they can only put an offer for 500 GOATS, it is better to error with a message than to change the amount.
+- Adding an upper timebound to a transaction can help the issuer ensure that their view of the world does not get out of sync.
+- Issuers can enforce additional fees by adding additional operations. For example, any transaction involving goats, will also send 0.1 GOAT to Boris’ account. 
+
+## Approval Server
+
+The transaction approval server receives a transaction, verifies compliance and either signs or rejects it.
```
[the remark this answers is not in the document]
Comment (tomerweller) [reply drafts/draft-0003.md]: OK. Given that this is repeated feedback and I'm not religious about it, let's unify the endpoints. I'll make a new revision with a unified endpoint ASAP.

**Comment by tomerweller** [comment]
@bartekn @michaeldowling I unified approval and validation into a single endpoint.

**Comment by bonen** [comment]
Great work, its  a very valuable addition to the protocol for our use case!
What I think might be required in _some_ scenarios; e.g. our case of delegated approval by a securities exchange is that the user is redirected from the wallet to the website of the exchange to visually see a confirmation. Our exchange (EU) must ensure that the user has been presented with honest information about exchange rate & orders etc. By showing a confirmation on the website of the exchange this can be ensured, it also enables a second factor (e.g. user signing in to exchange's website).

**Comment by tomerweller** [comment]
Thanks @bonen this is a good point. I'm wondering whether this should be part of the `pending` response, or perhaps we should add an `action_required` response. 

In any case, the response can include an `action_url` parameter that the issuer will generate on the fly.

**Comment by tomerweller** [comment]
@bonen I've added an `action_required` response. Check it out.

**Comment by abromberg** [comment]
This is really exciting. Thanks for all the hard work here. One question jumps out at me:

Some per-transaction restrictions could be determined programmatically rather than deferring directly to an issuer/exchange/etc. -- I'm thinking of e.g. 12g limits for certain equity securities. One example of this is for certain securities, there can be no more than 2,000 holders of record. 

Ideally, I'd think that a transaction in this category would be automatically approved if it didn't push the security over 2,000 holders of record and would be automatically denied if it did so. 

One way to solve this would be to have a programmatic endpoint controlled by the issuer that the protocol refers to for an answer. But ideally this could be done 100% within the protocol given that (I believe) all the information required to make that determination is there.

Does that make sense? Any thoughts?

**Comment by tomerweller** [comment]
Thanks for the feedback @abromberg 

You're right. There are many (if not an infinite amount of) constraints/restrictions that could be enforced just by looking at the current state of the ledger. To name a few: min/max account holders, a plethora of velocity limits (trades per hour, payments per day) and various fee structures (send 0.1 pig to the pigfarm whenever trading a goat) 

Given that the Stellar protocol doesn't currently support arbitrary custom logic (solidity style), we're left with two choices: (a) start integrating each and every possible constraint into the protocol or (b) defer to off-chain approval services.

This an attempt to standardize (b).

**Comment by capybaraz** [comment]
I propose we alter the name of the protocol from "Regulated Assets" to "Conditional Transfer Assets." There are numerous non-regulated use cases for this protocol -- I'd hate for the name of the protocol to pigeonhole its use case mainly for regulated assets.

**Comment by sui77** [comment]
I'm still quite new to stellar and I'm not sure if I completely got the whole concept. Basically this is not going to be an automagic stellar core feature but a standardized way for wallets to request a second signature from an approval server, right? So in order to utilize this I would need to...

- Set my issuer accounts auth_required flag=true (to prevent people from using non-multisig-accounts and thus be able to bypass this approval process)
- Somehow get users to add my approval server as a signer to their accounts
- Implement the approval server that would co-sign or decline user transactions
- Use a wallet (or implement my own) that would support this flow, i.e. request co-signing from the approval server before publishing the transaction to the network.

**Comment by sui77** [comment]
@swidnikk 
Revised example: There is a requirement to send a fee to a defined 3rd account with every payment but the transaction to be signed contains none or a wrong calculated fee. Approval service "proposes" a new acceptable transaction with correct fees. 

I agree that this has some potential for abuse since it will be challenging to clearly show the user the difference to the original tx. It could try to quietly add another signer or change any desastrous future account options that we are not even aware of today. On the other hand the user already trusts the issuer cause they wouldn't use that asset otherwise. However... besides I think this is also a bit YAGNI. I would probaly never bother implementing it in an approval service but just respond with a rejected + reason.

**Comment by swidnikk** [comment]
@sui77 Thanks for the example, below is an update to include that. I could envision it useful from a UX perspective if taken advantage of. If a service rejects transactions with some obtuse error message—though could be frustrating. On a personal note, with SWIFT transfers this has happened to me where a transaction failed but no one could tell me why (got to be pretty expensive =)

image

**Comment by tomerweller** [comment]
Thanks @swidnikk. This looks great. The `PENDING` response is intended to give the issuer more time to perform checks, maybe even get manual authorization. This is a bit different from how you describe it. 

How would you feel about writing some more information around this example and publishing it somewhere (blog, gist)? This would fit well in an external links section.

**Comment by johansten** [comment]
Moving things here since #135 got closed.

How does this work with state channels? I suspect it doesn't.

**Pull request merged**